File: Level3/단어변환.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;
int answer = 987654321;

void dfs(string current, string target, vector<string>& w, int index, vector<bool> check)
{
    if(current == target)
    {
        int ans = 0;
        for(int i = 0; i < w.size(); i++)
            if(check[i] == true)
                ans++;
        if(ans < answer)
            answer = ans;
        return;
    }
    
    
    
    for(int i = 0; i < w.size(); i++)
    {
        int count = 0;
        if(check[i] == false)
        {
            for(int j = 0; j < w[i].size(); j++)
                if(current[j] == w[i][j])
                    count++;
            if(count == current.size() - 1)
            {
                check[i] = true;
                dfs(w[i], target, w, i + 1, check);
                check[i] = false;
            }
        }
    }
}

int solution(string begin, string target, vector<string> words) 
{   
    if(find(words.begin(), words.end(), target) == words.end())
        return 0;
    vector<bool> check(words.size(), false);
    dfs(begin, target, words, 0, check);
    return answer;
}

File: Level3/단어 변환.js
const countDiffer = (word, target) => {
    let count = 0;
    
    for(let i = 0; i < word.length; i++){
        if(word[i] !== target[i]){
            count++;
        }
    }
    
    return count === 1 ? true : false;
}

const solution = (begin, target, words) => {
    let answer = 0;
    const queue = [];
    const map = new Map();
    
    queue.push(begin);
    
    while(queue.length){
        let size = queue.length;
        
        while(size--){
            const currentWord = queue[0];
            
            if(currentWord === target){
                return answer;
            }
            
            queue.shift();
            
            for(let i = 0; i < words.length; i++){
                const word = words[i];
                
                if(currentWord === word){
                    continue;
                }
                
                if(!map.get(currentWord + word) && !map.get(word + currentWord) && countDiffer(currentWord, word)){
                    queue.push(word);
                    map.set(currentWord + word, true);
                    map.set(word + currentWord, true);
                }
            }
        }
        
        answer++;
    }
    
    return 0;
}